sql: DROP hangs after server restart

(I think this is a repeat of an earlier issue I have filed last year, but I can't seem to find the earlier issue)

Scenario:
1. start 1-node cluster
2. start client
3. issue CREATE TABLE then SELECT on the table
4. shut down node with ctrl+c (once = graceful shutdown)
5. restart the node
6. issue DROP TABLE on the client

-> the DROP hangs, and the server logs get filled with

```
I170317 22:42:38.420390 316 sql/lease.go:301  publish (count leases): descID=110 name=blah version=1 count=1
I170317 22:42:38.596254 316 sql/lease.go:301  publish (count leases): descID=110 name=blah version=1 count=1
I170317 22:42:38.817968 316 sql/lease.go:301  publish (count leases): descID=110 name=blah version=1 count=1
I170317 22:42:39.050036 316 sql/lease.go:301  publish (count leases): descID=110 name=blah version=1 count=1
I170317 22:42:39.230746 316 sql/lease.go:301  publish (count leases): descID=110 name=blah version=1 count=1
I170317 22:42:39.407383 316 sql/lease.go:301  publish (count leases): descID=110 name=blah version=1 count=1
I170317 22:42:39.628070 316 sql/lease.go:301  publish (count leases): descID=110 name=blah version=1 count=1
```

cc @vivekmenezes

The node attempts to delete the lease but fails to send the RPC with
```
sql/lease.go:262  error releasing lease "51(\"t\") ver=1:1490214459728910000": failed to send RPC: sending to all 1 replicas failed; last error: node unavailable; try another peer
```
Will investigate further.

This is a single node issue. I'm sure you won't see it on a three node
cluster. We seem to have a situation where a node wants to drain and
therefore doesn't serve requests and yet it needs to serve requests in
order to drain. Perhaps when a nodes comes back up it also deletes old
leases

On Wed, Mar 22, 2017, 4:35 PM Alfonso Subiotto Marqués <
notifications@github.com> wrote:

> The node attempts to delete the lease but fails to send the RPC with
>
> sql/lease.go:262  error releasing lease "51(\"t\") ver=1:1490214459728910000": failed to send RPC: sending to all 1 replicas failed; last error: node unavailable; try another peer
>
> Will investigate further.
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/issues/14246#issuecomment-288530498>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBH9dUAkaAG1UXW7FJit1oqHeC70Uks5roYX_gaJpZM4MhLYf>
> .
>

--- a/pkg/cli/interactive_tests/test_server_restart.tcl
+++ b/pkg/cli/interactive_tests/test_server_restart.tcl
@@ -1,0 +1,22 @@
+#! /usr/bin/env expect -f
+#
+source [file join [file dirname $argv0] common.tcl]
+
+start_server $argv
+
+spawn /bin/bash
+send "PS1=':''/# '\r"
+eexpect ":/# "
+
+# Test that dropping a table a node held a lease on before a restart doesn't
+# hang. We use a SELECT statement to acquire a lease on the table.
+send "$argv sql -e \"create database t; create table t.t (x INT); select * from t.t;\"\r"
+eexpect "(0 rows)"
+
+stop_server $argv
+start_server $argv
+
+send "$argv sql -e \"drop table t.t;\"\r"
+eexpect "DROP TABLE"
+
+stop_server $argv

--- a/pkg/sql/lease.go
+++ b/pkg/sql/lease.go
@@ -660,7 +660,7 @@ func (t *tableState) acquireFreshestFromStoreLocked(
 
 // releaseInactiveLeases releases the leases in t.active.data with refcount 0.
 // t.mu must be locked.
-func (t *tableState) releaseInactiveLeases(store LeaseStore) {
+func (t *tableState) releaseInactiveLeases(m *LeaseManager) {
 	// A copy of t.active.data must be made since t.active.data will be changed
 	// by `removeLease`.
 	for _, lease := range append([]*LeaseState(nil), t.active.data...) {
@@ -668,7 +668,7 @@ func (t *tableState) releaseInactiveLeases(store LeaseStore) {
 			lease.mu.Lock()
 			defer lease.mu.Unlock()
 			if lease.refcount == 0 {
-				t.removeLease(lease, store)
+				t.removeLease(lease, m)
 			}
 		}()
 	}
@@ -767,20 +767,28 @@ func (t *tableState) release(lease *LeaseState, m *LeaseManager) error {
 		return s.released
 	}
 	if decRefcount(s) {
-		t.removeLease(s, m.LeaseStore)
+		t.removeLease(s, m)
 	}
 	return nil
 }
 
 // t.mu needs to be locked.
-func (t *tableState) removeLease(lease *LeaseState, store LeaseStore) {
+func (t *tableState) removeLease(lease *LeaseState, m *LeaseManager) {
 	t.active.remove(lease)
 	t.tableNameCache.remove(lease)
+
+	ctx := context.TODO()
+	if m.isDraining() {
+		// Release synchronously to guarantee release before exiting.
+		m.LeaseStore.Release(ctx, t.stopper, lease)
+		return
+	}
+
 	// Release to the store asynchronously, without the tableState lock.
-	if err := t.stopper.RunAsyncTask(context.TODO(), func(ctx context.Context) {
-		store.Release(ctx, t.stopper, lease)
+	if err := t.stopper.RunAsyncTask(ctx, func(ctx context.Context) {
+		m.LeaseStore.Release(ctx, t.stopper, lease)
 	}); err != nil {
-		log.Warningf(context.TODO(), "error: %s, not releasing lease: %q", err, lease)
+		log.Warningf(ctx, "error: %s, not releasing lease: %q", err, lease)
 	}
 }
 
@@ -823,7 +831,7 @@ func (t *tableState) purgeOldLeases(
 			if deleted {
 				t.deleted = true
 			}
-			t.releaseInactiveLeases(m.LeaseStore)
+			t.releaseInactiveLeases(m)
 			return nil
 		}
 		return err
@@ -1195,7 +1203,7 @@ func (m *LeaseManager) SetDraining(drain bool) {
 	defer m.mu.Unlock()
 	for _, t := range m.mu.tables {
 		t.mu.Lock()
-		t.releaseInactiveLeases(m.LeaseStore)
+		t.releaseInactiveLeases(m)
 		t.mu.Unlock()
 	}
 }

